src/gadget/sha256.rs

```diff
@@ -0,0 +1,274 @@
+
+    /// Places the SHA-256 IV in the circuit.
+    fn initialization_vector(
+        layer: &mut impl LayoutStrategy<Self>,
+        iv: &[u32; STATE],
+    ) -> Result<State<Self>, Error>;
+
+    /// Constructs the message schedule for a block of bits.
+    ///
+    /// Panics if `input.len()` is incorrect for this implementation.
+    fn message_schedule(
+        layer: &mut impl LayoutStrategy<Self>,
+        input: [Self::BlockWord; BLOCK_SIZE],
+    ) -> Result<[Self::MessageWord; ROUNDS], Error>;
+
+    /// Constrains `H′ = H + Ch(E,F,G) + Σ_1(E) + K_t + W_t`.
+    fn h_prime(
+        layer: &mut impl LayoutStrategy<Self>,
+        e: &Self::EFGH,
+        f: &Self::EFGH,
+        g: &Self::EFGH,
+        h: Self::EFGH,
+        k_t: u32,
+        w_t: &Self::MessageWord,
+    ) -> Result<Self::HPrime, Error>;
+
+    /// Constrains `A_{i+1} = H'_i + Maj(A_i,B_i,C_i) + Σ_0(A_i)`.
+    fn a_new(
+        layer: &mut impl LayoutStrategy<Self>,
+        a: &Self::ABCD,
+        b: &Self::ABCD,
+        c: &Self::ABCD,
+        h_prime: &Self::HPrime,
+    ) -> Result<Self::ABCD, Error>;
+
+    /// Constrains `E_{i+1} = D_i + H'_i`.
+    fn e_new(
+        layer: &mut impl LayoutStrategy<Self>,
+        d: Self::ABCD,
+        h_prime: &Self::HPrime,
+    ) -> Result<Self::EFGH, Error>;
+
+    /// Adds two 32-bit words of type [`Self::ABCD`].
+    fn add_abcd(
+        layer: &mut impl LayoutStrategy<Self>,
+        lhs: &Self::ABCD,
+        rhs: &Self::ABCD,
+    ) -> Result<Self::ABCD, Error>;
+
+    /// Adds two 32-bit words of type [`Self::EFGH`].
+    fn add_efgh(
+        layer: &mut impl LayoutStrategy<Self>,
+        lhs: &Self::EFGH,
+        rhs: &Self::EFGH,
+    ) -> Result<Self::EFGH, Error>;
+}
+
+/// Implementation of the compression function for SHA-256.
+#[derive(Debug)]
+pub struct Sha256Compress<CS: Sha256CompressInstructions> {
+    _cs: PhantomData<CS>,
+}
+
+impl<CS: Sha256CompressInstructions> Sha256Compress<CS> {
+    /// Starting from the given intial state, processes a block of input and returns the
+    /// final state.
+    fn process(
+        layer: &mut impl LayoutStrategy<CS>,
+        initial_state: &State<CS>,
+        input: [CS::BlockWord; 16],
+    ) -> Result<State<CS>, Error> {
+        let mut a = initial_state.h_0.clone();
+        let mut b = initial_state.h_1.clone();
+        let mut c = initial_state.h_2.clone();
+        let mut d = initial_state.h_3.clone();
+        let mut e = initial_state.h_4.clone();
+        let mut f = initial_state.h_5.clone();
+        let mut g = initial_state.h_6.clone();
+        let mut h = initial_state.h_7.clone();
+
+        let w = CS::message_schedule(layer, input)?;
+
+        for (&k_t, w_t) in ROUND_CONSTANTS.iter().zip(w.iter()) {
+            let h_prime = CS::h_prime(layer, &e, &f, &g, h, k_t, w_t)?;
+
+            let a_new = CS::a_new(layer, &a, &b, &c, &h_prime)?;
+            let e_new = CS::e_new(layer, d, &h_prime)?;
```

Comment:
A chip may have an implementation of `h_prime`, `a_new`, and `e_new` that use relative cell references to pass variables between rows. The permutation arguments may not cover all advice columns, and so relative references may be required for efficiency. The layouter is supposed to be a general algorithm; the chip could require certain capabilities from it, but it's not specific to this chip. And the gadget could call the instruction methods in any order.

How will the layouter construct a layout that wires up the relative references correctly?

Reply:
One possibility is that you can only use relative references within an instruction, but I think this is too restrictive. Consider a scalar multiplication or Sinsemilla gadget: we know that there's a natural ordering of the steps of Sinsemilla or of a particular scalar multiplication algorithm. The problem is to design an API that separates concerns between the gadget and the chip, while constraining the gadget to make calls in the "right" order to avoid the layouter having to solve an unnecessary hard layout problem. One way to do it is using session types: the chip defines an API that enforces a particular ordering of calls from the gadget. But I have a feeling we're missing something, since this seems overcomplicated. In the cases we want to immediately support (scalar mul, Sinsemilla, Rescue and Poseidon), the required ordering is just a linear sequence of steps. This suggests a more functional API: we construct a vector of steps for a gadget and pass that to the layouter all at once. Then the layouter has a much easier job. In general it might need to lay out instructions from different gadgets in parallel, but we could reasonably require (at least for now) that the permutation argument covers enough columns to allow loading data in and out of gadget blocks that are laid out in parallel.

Reply:
On further consideration, I think we cannot meaningfully separate concerns between the gadget and the chip without letting the chip essentially implement the whole high-level algorithm.

I started off by trying to change the Chip API in the simplest way I could think of to avoid the call ordering problem described above:
```
impl<Chip: Sha256CompressInstructions> Sha256Compress<Chip> {
    fn process(
        layer: &mut impl LayoutStrategy<Chip>,
        initial_state: &State<Chip>,
        input: [Chip::BlockWord; 16],
    ) -> Result<State<Chip>, Error> {
        let mut state = initial_state.clone();
        let w = Chip::message_schedule(layer, input)?;
        for (&k_t, w_t) in ROUND_CONSTANTS.iter().zip(w.iter()) {
            state = Chip::round(&state, k_t, w_t)?;
        }
        Chip::feedforward(&state, &initial_state)?;
}
```
with the assumption that `round` will be called in the obvious order.

But at this point we're not really gaining anything from separating out the `message_schedule`, `round`, and `feedforward` functions. We might as well just define a `compress` instruction that does the whole thing.

Reply:
@daira and I discussed this in a pairing, and we discussed a variety of issues for which I now have notes 😄 For this PR, we will take the approach of defining a single `compress` instruction. My decision to agree on this was in part influenced by seeing that the Intel SHA extensions define an instruction that operates on two compression rounds simultaneously, which is less granular than the instructions we'd defined here.

Reply:
I would really like to look at how we might construct this API starting from the high-level view of a gadget/chip/whateveryouwanttocallit as a transformation on tables, and then from there deconstruct to the smaller transformations on tables that need to be composed to achieve the larger one. I really think that designing this API imperatively is muddying the waters. One can always implement a declarative API in terms of mutation.

Reply:
I agree with @nuttycom that it is simpler and more general to start from a declarative API design.

In #93 I sketched out a simple layout algorithm that should suffice for the circuits we need in the short term. This clarifies that all the input a layout algorithm needs is the block shapes (where relative references are only allowed within a block). So I claim that a gadget should just output a set of blocks. There's no requirement that it use an imperative API to do so.